src/join.sol

```diff
@@ -502,3 +502,76 @@ contract GemJoin7 is LibNote {
         gem.transfer(guy, wad);
     }
 }
+
+// GemJoin8
+// For a token that has a lower precision than 18, has decimals and it is upgradable (like GUSD)
+
+interface GemLike8 {
+    function decimals() external view returns (uint8);
+    function transfer(address,uint) external returns (bool);
+    function transferFrom(address,address,uint) external returns (bool);
+    function erc20Impl() external view returns (address);
+}
+
+contract GemJoin8 is LibNote {
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike  public vat;
+    bytes32  public ilk;
+    GemLike8 public gem;
+    bytes    public funcSig;
+    uint     public dec;
+    uint     public live;  // Access Flag
+
+    mapping (address => uint256) public implementations;
+
+    constructor(address vat_, bytes32 ilk_, address gem_, string memory func_) public {
+        gem = GemLike8(gem_);
+        dec = gem.decimals();
+        require(dec < 18, "GemJoin8/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        funcSig = abi.encodeWithSignature(func_);
```

Comment:
Interesting solution for alternate implementation checks. This constructor could benefit from some documentation of the parameters to ensure that it's the implementation call that's getting passed here, and not, say, a `transfer()` function. Using a name like `implFunc_` might also help clarify.